### ISSUE TITLE
Add eID python types

### ISSUE THREAD

**Issue by m273d15** [swp-fu-eid/eid-fu-swp#51]
Types to implement (see [eID-Server specification](https://www.bsi.bund.de/SharedDocs/Downloads/DE/BSI/Publikationen/TechnischeRichtlinien/TR03130/TR-03130_TR-eID-Server_Part1.pdf;jsessionid=BBBF6A846DFA4F960CA82C7ECA84313D.2_cid351?__blob=publicationFile&v=3)):

- [x] SessionType [Armagetron]
   - [x] eid-types/sessionid.py
   - [x] tests/sessionid.py
   - [x] eid-types/type_check.py
- [x] RestrictedIDType
   - [x] eid-types/restrictedid.py
   - [x] tests/restrictedid.py
   - [x] eid-types/type_check.py
      - [x] check(..., RestrictedID)
      - [x] check(..., HexBinary)
      - [x] check_is_string_matching()
- [x] PersonalDataType [larissazech]
   - [x] eid-types/personaldata.py
      - [x] ICAOCountry.check_icao_country()
      - [x] GeneralPlace.check_general_place()
   - [x] tests/personaldata.py
   - [x] eid-types/type_check.py
      - [x] check_is_string_matching()
      - [x] check(..., DocumentType)
      - [x] check_general_date()
      - [x] check_str()
      - [x] check_restrictedID()
- [x] ICAOCountry
   - [x] eid-types/place.py
   - [x] tests/place.py
   - [x] eid-types/type_check.py
      - [x] check(..., ICAOCountry)
      - [x] check_str()
- [x] GeneralPlace
   - [x] eid-types/place.py
      - [x] Place.check_place()
   - [x] tests/place.py
   - [x] eid-types/type_check.py
      - [x] check(..., GeneralPlace)
      - [x] check_str()
- [x] PlaceType
   - [x] eid-types/place.py
      - [x] ICAOCountry.check_icao_country()
   - [x] tests/place.py
   - [x] eid-types/type_check.py
      - [x] check(..., Place)
      - [x] check_str()
- [x] OperationsSelectorType
   - [x] eid-types/operationsselector.py
   - [x] tests/operationsselector.py
   - [x] eid-types/type_check.py
      - [x] check_attributeSelection()
- [x] OperationsRequestorType
   - [x] eid-types/operationsrequestor.py
   - [x] tests/operationsrequestor.py
   - [x] eid-types/type_check.py
      - [x] check_attributeRequest()
- [x] OperationsResponderType
   - [x] eid-types/operationsresponder.py
   - [x] tests/operationsresponder.py
   - [x] eid-types/type_check.py
      - [x] check_attributeResponse()
- [x] AgeVerificationRequestType
   - [x] eid-types/ageverificationrequest.py
   - [x] tests/ageverificationrequest.py
   - [x] eid-types/type_check.py
      - [x] check_positive_int()
- [x] PlaceVerificationRequestType
   - [x] eid-types/placeverificationrequest.py
   - [x] tests/placeverificationrequest.py
   - [x] eid-types/type_check.py
      - [x] check_placeVerificationRequest()
- [x] VersionType
   - [x] eid-types/version.py
   - [x] tests/version.py
   - [x] eid-types/type_check.py
      - [x] check_int()
      - [x] check_str()
- [x] VerificationResultType
   - [x] eid-types/verificationresult.py
   - [x] tests/verificationresult.py
   - [x] eid-types/type_check.py
      - [x] check_bool()
- [x] PreSharedKeyType [m273d15]
   - [x] eid-types/.py
   - [x] tests/.py
   - [x] eid-types/type_check.py
- [x] GeneralDateType [nielsgroth]
   - [x] eid-types/generaldate.py
   - [x] tests/generaldate.py
   - [x] eid-types/type_check.py
      - [x] check_dateValue()
- [x] AttributeSelectionType
   - [x] eid-types/attributeselection.py
   - [x] tests/attributeselection.py
   - [x] eid-types/type_check.py
      - [x] check_attributeSelection()
- [x] AttributeRequestType
   - [x] eid-types/attributerequest.py
   - [x] tests/attributerequest.py
   - [x] eid-types/type_check.py
      - [x] check_attributeRequest()
- [x] AttributeResponseType
   - [x] eid-types/attributeresponse.py
   - [x] tests/attributeresponse.py
   - [x] eid-types/type_check.py
      - [x] check_attributeResponse

**Comment by m273d15** [comment]
Implementing VersionType and starting to configure a test infrastructure for python scripts

**Comment by Armagetron** [comment]
Waiting for first commit. Then I will add SessionType.

**Comment by m273d15** [comment]
Implementing PlaceType and ICAOCountry

**Comment by larissazech** [comment]
Implementing VerificationResultType, AttributeSelectionType and OperationsSelectorType.

**Comment by m273d15** [comment]
implementing GeneralPlaceType

**Comment by larissazech** [comment]
Implementing AttributeRequestType and OperationsRequestorType.

**Comment by auvin** [comment]
Implementing AttributeResponseType

**Comment by auvin** [comment]
Some Types need Datatypes which are not described in the [eID-Server specification](https://www.bsi.bund.de/SharedDocs/Downloads/DE/BSI/Publikationen/TechnischeRichtlinien/TR03130/TR-03130_TR-eID-Server_Part1.pdf;jsessionid=BBBF6A846DFA4F960CA82C7ECA84313D.2_cid351?__blob=publicationFile&v=3) like AgeType or CommunityIDType. 
I think some of them are described in [TR-03127](https://www.bsi.bund.de/SharedDocs/Downloads/EN/BSI/Publications/TechGuidelines/TR03127/BSI-TR-03127_en.pdf?__blob=publicationFile&v=1) and we have to implement these as well.

**Comment by auvin** [comment]
>  The data types involve mostly complex data types defined in the XML Schema Definition (XSD). The corresponding XSD-File TR-03130eID-Server.xsd is attached to this guideline.

So in  the [TR-03130eID-Server.xsd](https://www.bsi.bund.de/SharedDocs/Downloads/DE/BSI/Publikationen/TechnischeRichtlinien/TR03130/TR-03130_TR-eID-Server_WSDL.null;jsessionid=95380C685A82E2F34C1D0E8F84A069CF.1_cid360?__blob=publicationFile&v=2) are the complex data types described.

**Comment by auvin** [comment]
The GeneralDateType is described as follows:  

```
<complexType name="GeneralDateType">
	<sequence>
		<element name="DateString">
			<simpleType>
				<restriction base="string">
					<whiteSpace value="preserve" />
					<pattern value="[0-9 ]{8}" />
				</restriction>
			</simpleType>
		</element>
		<element name="DateValue" type="date" minOccurs="0" />
	</sequence>
</complexType>
```

So the element `DateValue` has type date, which i think is reffering to the XSD type [xs:date](https://www.w3schools.com/xml/schema_dtypes_date.asp). Now i am not sure if ishould implement this datatype as a simple string, with an own date-class or with the build in date-object provided by python ([datetime.date](https://docs.python.org/2/library/datetime.html#datetime.date))?

Solution: Take the the build in classes/objects from python.

**Comment by Armagetron** [comment]
SessionID added

**Comment by Armagetron** [comment]
WSDLs are available in this branch: https://github.com/swp-fu-eid/eid-fu-swp/tree/eid_server_wsdl

**Comment by larissazech** [comment]
@nielsgroth and I are quite confused by the specifications of the GeneralDateType, or, more specifically, of the **DateString**. The definition in the eID-Server specification reads as follows:

> The data type GeneralDateType MUST contain the date representation read directly from the chip in the DateString element. An 8-place string SHALL be used here that MUST contain spaces as well as 
numerical values (0-9). It's transmitted in the yyyymmdd format (4 places for the year, 2 places for the 
month and 2 places for the day) as specified in [EAC] Part 2 Annex A Section 2: ASN.1 Definition. The optional element DateValue of the simple type date MUST only be present if the eID-Server was able to read a complete date from the eID Application.

and taken from the xsd (as already mentioned above):

> <complexType name="GeneralDateType">
	<sequence>
		<element name="DateString">
			<simpleType>
				<restriction base="string">
					<whiteSpace value="preserve" />
					<pattern value="[0-9 ]{8}" />
				</restriction>
			</simpleType>
		</element>
		<element name="DateValue" type="date" minOccurs="0" />
	</sequence>
</complexType>

Now, what format do we accept? Where do we accept whitespaces? Does the DateString really **must** contain whitespaces? Somewhere...? And do they count into the 8 places? And do we keep them?

@Armagetron @nils-wisiol feedback is greatly appreciated.

**Comment by Armagetron** [comment]
I understand the definition in a way that the date should have the form "20170825" but "2017 08 25" should be understood as well.

**Comment by larissazech** [comment]
All specified types seem to be implemented by now. 

@Armagetron @m273d15 @nielsgroth What should be done in preparation of the pull request?

**Comment by nils-wisiol** [comment]
I believe 

> MUST contain spaces as well as
> numerical values (0-9)

should be read as in: must contain (spaces and numerical values). So either is fine. Then I'd read

>  An 8-place string SHALL be used here

as in: it could be more or less than 8 chars.

Therefore, I agree with @Armagetron that, judging from the docs, `2017 08 25` should be accepted. `<pattern value="[0-9 ]{8}" />` however does not match `2017 08 25` (wtf?).

I propose, whenever we need to parse a date formatted like this, remove all whitespace before parsing. Whenever we render a date like this, do not include whitespace. That way we should somewhat follow the standard and be compatible to whoever else had to implement this.

**Comment by larissazech** [comment]
Implemented by #70. Closed.
